### PR TITLE
Bug 1588661 - Design for Webhooks

### PR DESCRIPTION
**Pull request by LissetCuevas**
@dklawren what it’s happening is:
When I run it like it is in this PR, when I open the webhook tab, it throws me an error that says: 

DBD::mysql::db selectall_arrayref failed: Unknown column 'event' in 'field list' [for Statement "SELECT id,name,url,user_id,event,product_id,component_id FROM webhook WHERE user_id = ?  ORDER BY id"] at /app/Bugzilla/DB.pm line 71.

The problem is that for some reason that I can’t find the webhook table isn’t been created with the event column.
And when I comment all the lines related to event column in all the files the extension works correctly. 

I review the code and I think all it’s declared correctly. But what I said in the meet is that first I built the extension for a table without the event column and when all was working I added all the event column. I think it could be the reason but I tried some ways to solve it if that were the case but didn’t solve the problem.

### CODE REVIEW

**Comment by LissetCuevas** [comment]
The changes were made and the feature flag added.

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl [hunk 1]
```diff
@@ -64,8 +80,7 @@ YAHOO.util.Event.onDOMReady(onRemoveChange);
 </script>
 
 <p>
-  It will be sent a POST request with all the information of the bugs that match with the events and filters selected to your URL.<br>
-  Documentation about webhooks is available <a href="">here</a>.
+  It will be sent a POST request with the information of the bugs that match with the events and filters selected to your URL.
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl]: It will be sent a POST request with the information of the [% terms.bugs %] that match with the events and filters selected to your URL.

File: extensions/Webhooks/template/en/default/hook/global/user-error-errors.html.tmpl [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  [% title = "Select at least one event" %]
+  You didn't select any event. Select at least one.
+
+[% ELSIF error == "define_a_name" %]
+  [% title = "Define a name" %]
+  You didn't define a name for your webhook. Define a name to identify your webhook.
+
+[% ELSIF error == "define_a_url" %]
+  [% title = "Define a URL" %]
+  You didn't define a URL for your webhook. Difine one.
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/hook/global/user-error-errors.html.tmpl:19]: s/Difine/Define/

**Comment by LissetCuevas** [comment]
I made the changes to standard Javascript. I think it's ready but if there's anything else please tell me.

**Comment by dklawren** [comment]
Warnings that need to be addressed from the log:
bmo.test_1      | Use of uninitialized value in numeric eq (==) at /app/extensions/Webhooks/Extension.pm line 97.
bmo.test_1      | Use of uninitialized value in numeric eq (==) at /app/extensions/Webhooks/Extension.pm line 100.
bmo.test_1      | Use of uninitialized value in numeric eq (==) at /app/extensions/Webhooks/Extension.pm line 103.

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+window.onload = function() {
+  onSelectComponent();
+  onSelectProduct();
+  onRemoveChange();
+};
+</script>
+
+<p>
+  It will be sent a POST request with the information of the [% terms.bugs %] that match with the events and filters selected to your URL.<br>
+  Documentation about webhooks is available <a href="">here</a>.
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl:77]: This should be a page.cgi template and point to it.  

Example template
extensions/BMO/template/en/default/pages/etiquette.html.tmpl

The new one would be:
extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl

We will add any user docs there and access the page using `page.cgi?id=webhooks.html`

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+<tr>
+  <th align="right">URL:</th>
+  <td colspan="2"><input type="text" name="url" id="url" maxlength="64"></td>
+</tr>
+</table>
+<h4>Events</h4>
+<p>Select the events you want to receive.</p>
+<p>
+  <input type="checkbox" id="create_event" name="create_event" value="1">
+  <label>Create</label>
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl:94]:
```html
<label for="create_event">Create</label>
```

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+<br>
+  <input type="checkbox" id="change_event" name="change_event" value="1">
+  <label>Change</label>
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl:97]:
```html
<label for="change_event">Change</label>
```

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        [%~ product.name FILTER html %]</option>
+    [% END %]
+    </select>
+  </td>
+</tr>
+<tr>
+  <th align="right">Component:</th>
+  <td colspan="2">
+    <select name="add_component" id="component" onChange="onSelectComponent()>
+      <option value="">__Any__</option>
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl:119]: The Any option should be selected by default in the new filters.

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+  [% END %]
+  </tbody>
+  </table>
+  <br>
+  <input id="remove" type="submit" value="Remove Selected">
+
+[% ELSE %]
+
+  <p>
+    <i>You do not have any webhook.</i>
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl:179]: You do not have any webhooks.

File: extensions/Webhooks/template/en/default/hook/global/user-error-errors.html.tmpl [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  [% title = "Select at least one event" %]
+  You didn't select any event. Select at least one.
+
+[% ELSIF error == "define_a_name" %]
+  [% title = "Define a name" %]
+  You didn't define a name for your webhook. Define a name to identify your webhook.
+
+[% ELSIF error == "define_a_url" %]
+  [% title = "Define a URL" %]
+  You didn't define a URL for your webhook. Define one.
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/hook/global/user-error-errors.html.tmpl:19]: Small request: Prefix all errror names with the extension name so avoid any naming conflicts.

webhooks_select_event
webhooks_define_name
webhooks_define_url

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl (hunk 3 above)
Comment (LissetCuevas) [reply extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl]: I added the page template and the link but the page content isn't complete. I will work on that after I finish the Push Connector.

**Comment by LissetCuevas** [comment]
@emceeaich We have a first version (maybe almost first) of Webhooks Feature!! 🎉

To test it, I use [Ngrok](https://ngrok.com/) and a very easy mojolicious app: 

```
use Mojolicious::Lite;

post '/test' => sub {
  my $c    = shift;
  $c->render(data => '', status => 200);
  app->log->debug('I got a message! See it in http://127.0.0.1:4040 :)');
};

app->start;
```

I was guided by what github suggest (https://developer.github.com/webhooks/configuring/). I'm not sure if a should list something similar in the Webhooks specifications, like "Testing Webhooks".

I'll wait for review, and while, I have pending Documentation Page and work on my CV, so I work on that on Monday.

**Comment by emceeaich** [comment]
An example would be good, but probably something using flask as the example app.

**Comment by LissetCuevas** [comment]
> An example would be good, but probably something using flask as the example app.
Use flask app instead mojolicious app?

**Comment by LissetCuevas** [comment]
@emceeaich The last commit is the Documentation Page.

**Comment by dklawren** [comment]
t/009bugwords.t .............................. 1/725 
#   Failed test '/app/extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl contains invalid bare words (e.g. 'bug') --WARNING'
#   at t/009bugwords.t line 72.
#

**Comment by LissetCuevas** [comment]
> t/009bugwords.t .............................. 1/725
> Failed test '/app/extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl contains invalid bare words (e.g. 'bug') --WARNING'
> at t/009bugwords.t line 72.

@dklawren This problem is because it's used the word bug, Should I write the page without use this word or is there another way to solve the problem? because I think we need this words in the Documentation Page to understand better it :/

File: extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            }
+            "
+%]
+
+<h1>Bugzilla Webhooks</h1>
+
+<p>A webhook is a custom callback defined by events. Is triggered when those events happen, and a POST
+request is sent to a defined URL.</p>
+
+<p>In the case of Bugzilla, a WebHook can be triggered by a change to, or creation of a bug, the
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl:33]: Replace words Bugzilla with [% terms.Bugzilla %]
Replace words bug with [% terms.bug %]
Replace words Bug with [% terms.Bug %]

File: extensions/Webhooks/Extension.pm [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+      },
+      component_id => {
+        TYPE       => 'INT2',
+        NOTNULL    => 0,
+        REFERENCES => {TABLE => 'components', COLUMN => 'id', DELETE => 'CASCADE',}
+      }
+    ],
+  };
+}
+
```
Comment (dklawren) [inline extensions/Webhooks/Extension.pm:56]: Let's add a UNIQUE INDEX for the user_id and name columns. That way a user cannot have more than one webhook with the same name.

```
INDEXES => [
      webhooks_userid_name_idx => {FIELDS => ['user_id', 'name'], TYPE => 'UNIQUE',},
    ],
```

**Comment by LissetCuevas** [comment]
> For extensions/Push/lib/Config.pm I am thinking that for any connector that is of type webhook we just omit completely from the Admin UI page. Once more and more people add their own hooks, the admin page is going to get very long and is not necessary for the admin to see anyway. We can add it back or come up with a different page specifically for administration of webhooks. I can always disable it from the command line anyway if we need to turn off a misbehaving webhook. I am open to other opinions on this.

I think have a different page specifically for administration of webhooks is a good option, because I asked to Emma and Glob about that in a meeting and they said that it will be useful for the administrators see the configurations, so we can omit them from push config panel and maybe make a new page with a table with all the webhooks configurations just to see them 🤔

**Comment by dklawren** [comment]
> > For extensions/Push/lib/Config.pm I am thinking that for any connector that is of type webhook we just omit completely from the Admin UI page. Once more and more people add their own hooks, the admin page is going to get very long and is not necessary for the admin to see anyway. We can add it back or come up with a different page specifically for administration of webhooks. I can always disable it from the command line anyway if we need to turn off a misbehaving webhook. I am open to other opinions on this.
> 
> I think have a different page specifically for administration of webhooks is a good option, because I asked to Emma and Glob about that in a meeting and they said that it will be useful for the administrators see the configurations, so we can omit them from push config panel and maybe make a new page with a table with all the webhooks configurations just to see them thinking

Sounds good. I can also speak to them to see why else they might want to see them on the normal push connector admin page. I just feel like it needs a different page all its own as the list will get long over time and the way the push connector admin page is laid out, it will be very cluttered. The only reason personally I would want to maybe see all of the webhooks in a page as a list, is if we need to manually disable one that is not working and looping forever. We would need a drop down or checkbox to temporarily disable the webhook and then notify the owner of the problem.

So basically the normal push connector admin page could include a link near the bottom to the special page for administering webhooks if webhook support is enabled.

**Comment by LissetCuevas** [comment]
> Sounds good. I can also speak to them to see why else they might want to see them on the normal push connector admin page. I just feel like it needs a different page all its own as the list will get long over time and the way the push connector admin page is laid out, it will be very cluttered. The only reason personally I would want to maybe see all of the webhooks in a page as a list, is if we need to manually disable one that is not working and looping forever. We would need a drop down or checkbox to temporarily disable the webhook and then notify the owner of the problem.
> 
> So basically the normal push connector admin page could include a link near the bottom to the special page for administering webhooks if webhook support is enabled.

Yes, it's a good idea. I'll start removing webhooks from config panel while we know what Emma and Glob think.
About the case of a webhook in a loop I was thinking that we can send an email or a notification to the user and/or the admin after certain number of attempts to inform about the error and they could check it, and maybe also automatically disable the webhook.

**Comment by dklawren** [comment]
> > Sounds good. I can also speak to them to see why else they might want to see them on the normal push connector admin page. I just feel like it needs a different page all its own as the list will get long over time and the way the push connector admin page is laid out, it will be very cluttered. The only reason personally I would want to maybe see all of the webhooks in a page as a list, is if we need to manually disable one that is not working and looping forever. We would need a drop down or checkbox to temporarily disable the webhook and then notify the owner of the problem.
> > So basically the normal push connector admin page could include a link near the bottom to the special page for administering webhooks if webhook support is enabled.
> 
> Yes, it's a good idea. I'll start removing webhooks from config panel while we know what Emma and Glob think.
> About the case of a webhook in a loop I was thinking that we can send an email or a notification to the user and/or the admin after certain number of attempts to inform about the error and they could check it, and maybe also automatically disable the webhook.

Sounds great. We can do the email to the user thing as a next release feature. AS long as admin can turn it off for now that would good enough. Our system administrators monitor the push log and get alerted if the errors get to be too high so they will let us know when to check the logs.

**Comment by LissetCuevas** [comment]
> Sounds great. We can do the email to the user thing as a next release feature. AS long as admin can turn it off for now that would good enough. Our system administrators monitor the push log and get alerted if the errors get to be too high so they will let us know when to check the logs.

Great, so I'll add the page to see the webhooks too :)

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+  Your webhooks:
+</h3>
+
+[% IF webhooks.size %]
+
+  <table id="webhooks_table" class="standard">
+  <thead>
+    <tr>
+      <th>&nbsp;</th>
+      <th>ID</th>
```
[the remark this answers is not in the document]
Comment (LissetCuevas) [reply extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl:132]: I added the ID thinking that the queues are named with the ID so if there's a problem and a user ask for it can said which ID has and the administrator can easily find it. Do you think could be the case or I remove it?

File: extensions/Push/lib/Connector/Webhook.pm [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+      required => 0,
+    },
+  );
+}
+
+sub new {
+  my ($class,$webhook_id) = @_;
+  my $self = {};
+  bless($self, $class);
+  ($self->{name}) = 'Webhook_' . $webhook_id;
```
[the remark this answers is not in the document]
Comment (LissetCuevas) [reply extensions/Push/lib/Connector/Webhook.pm]: I think we can't do this, because the connector is initialized with the webhook_id, from there get the object webhook to configure it (in the load_config) so if we don't pass it we don't have the self->id, and in the name is named with that when is created and as isn't still configure when is named, it don't have the self->id so we need to pass it too.

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl [hunk 13]
```diff
@@ -98,7 +98,7 @@ window.onload = function() {
   </td>
 </tr>
 <tr>
-  <th>Component:</th>
+  <th align="right">Component:</th>
   <td>
     <select name="add_component" id="component">
       <option value="">__Any__</option>
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl:104]: Change __Any__ to be Any to be consistent everywhere.

File: extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl [hunk 14]
```diff
@@ -54,23 +54,23 @@ web application.</p>
 
     <h4>Events</h4>
     <p>The [% terms.bug %] events that will trigger your new webhook</p>
-      <ul>
-        <li><b>create</b>: when a new [% terms.bug %] is created.</li>
-        <li><b>change</b>: when an existing [% terms.bug %] has some change.</li>
+      <ul>[% terms.bug %][% terms.bug %]
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl:57]: Remove the extra [% terms.bug %][% terms.bug %]

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl [hunk 15]
```diff
@@ -66,29 +66,29 @@ window.onload = function() {
...
 <h4>Events</h4>
 <p>Select the events you want to receive.</p>
 <p>
   <input type="checkbox" id="create_event" name="create_event" value="1">
-  <label for="create_event">Create</label>
+  <label for="create_event">When a new bug is created</label>
 <br>
   <input type="checkbox" id="change_event" name="change_event" value="1">
-  <label for="change_event">Change</label>
+  <label for="change_event">When an existing bug is modified</label>
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl:83]: [% terms.bug %]

File: extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl [hunk 16]
```diff
@@ -66,29 +66,29 @@ window.onload = function() {
...
-  <td colspan="2"><input type="text" name="url" id="url" maxlength="64"></td>
+  <td><input type="text" name="url" id="url" maxlength="64"></td>
 </tr>
 </table>
 <h4>Events</h4>
 <p>Select the events you want to receive.</p>
 <p>
   <input type="checkbox" id="create_event" name="create_event" value="1">
-  <label for="create_event">Create</label>
+  <label for="create_event">When a new bug is created</label>
```
Comment (dklawren) [inline extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl:80]: [% terms.bug %]

**Comment by dklawren** [comment]
Also while you are working on the admin related screens, lets remove any sections to do with webhooks from the push_queues.html screen as well that have (0) entries.

image

If the webhook has one or more items queued or failures then we display otherwise they are skipped. Reason is this page (like the config page) will get rather long once we have a lot of webhooks in the system.

Sorry for these last minute changes :smile:

File: extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl [hunk 17]
```diff
@@ -0,0 +1,190 @@
+[%# This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (emceeaich) [inline extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl:1]: This should probably be moved into /docs as an `.rst` file.

File: extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+web application.</p>
+
+<p>Examples of [% terms.Bugzilla %] webhooks could include:</p>
+<ul>
+  <li>Updating a copy of the [% terms.Bugzilla %] [% terms.bug %] in another system such as Jira</li>
+  <li>Sending a message to a chat server such as Matrix or Slack</li>
+</ul>
+
+<h2>Creating a webhook</h2>
+<p>To create a webhook you just need to follow the next steps:</p>
```
Comment (emceeaich) [inline extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl:44]: remove `just need to`.

File: extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+<ol>
+  <li>Access to your [% terms.Bugzilla %] Account.</li>
+  <li>Go to your Preferences Panel > Webhooks.</li>
+  <li>Fill out all of the parameters:</li>
+    <h4>Name</h4>
+    <p>A name for you to recognize the webhook. (e.g. “My Jira webhook”)</p>
```
Comment (emceeaich) [inline extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl:50]: `A name for the webhook, which should be descriptive (e.g. `Jira Webhook for New & Updated Bugs in Core::Graphics`

File: extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+
+    <h4>URL</h4>
+    <p>A valid URL where the callback will be sent.</p>
```
Comment (emceeaich) [inline extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl:53]: `The URL which will receive and process the webhook.`

Should this list what responses are expected from the webhook URL?

**Comment by dklawren** [comment]
The new changes look good. Just need to fix the failing sanity test

t/009bugwords.t .............................. 1/725 
#   Failed test '/app/extensions/Webhooks/template/en/default/account/prefs/webhooks.html.tmpl contains invalid bare words (e.g. 'bug') --WARNING'
#   at t/009bugwords.t line 72.
#

File: extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl (hunk 17 above)
Comment (dklawren) [reply extensions/Webhooks/template/en/default/pages/webhooks.html.tmpl:1]: Not a bad idea but given the time frame let's file it as a post MVP enhancement to be fixed later.

**Comment by LissetCuevas** [comment]
Thanks a lot for all your reviews and mentorship! :smile: 
I think that also we could file a bug to allow the users to enable and disable their webhooks. If you want, I can file the bugs and start working on them.

**Comment by dklawren** [comment]
> Thanks a lot for all your reviews and mentorship! smile
> I think that also we could file a bug to allow the users to enable and disable their webhooks. If you want, I can file the bugs and start working on them.

Oh yeah forgot. Let's file that one as well. Absolutely if you want to file and work on as much as you have time left that would be great. Not sure what the final requirements are for the outreachy program as far as what you have do turn in but let me know what I can do. Thanks.

**Comment by LissetCuevas** [comment]
Great, I'll do it. The internship ends until August 18th so I can work on some bugs, and anyway I want to continue contributing later :) Thanks, if there's anything I'll tell you.
